### PR TITLE
allow relative path for configuration file

### PR DESCRIPTION
**Pull request by majestrate**

### CODE REVIEW

**Comment by majestrate** [comment]
fixes #1430

File: llarp/config/config.cpp [hunk 1]
```diff
@@ -1132,10 +1132,12 @@ namespace llarp
     if (ec)
       throw std::runtime_error(stringify("filesystem error: ", ec));
 
+    const auto parent = confFile.parent_path();
+
     // create parent dir if it doesn't exist
-    if (not fs::exists(confFile.parent_path(), ec))
+    if ((not parent.empty()) and (not fs::exists(parent, ec)))
```
Comment (tewinget) [inline llarp/config/config.cpp]: Since we're passing `basedir` in here as `defaultDataDir`, it's weird to use `parent_path` again here.

Comment (tewinget) [reply llarp/config/config.cpp]: Also the parameter `defaultDataDir` here should probably be renamed; suggestion: `dataDir`.

File: llarp/config/config.cpp [hunk 2]
```diff
@@ -107,6 +107,8 @@ namespace llarp
             "private keys.",
         },
         [this](fs::path arg) {
+          if (arg.empty())
```
Comment (tewinget) [inline llarp/config/config.cpp:110]: imo this should be a fatal error.  The case where the `data-dir` setting is present in the config file but is an empty string is almost certainly user error and should be treated as such.

Comment (majestrate) [reply llarp/config/config.cpp:110]: good point

File: llarp/config/config.cpp [hunk 3]
```diff
@@ -1143,9 +1145,13 @@ namespace llarp
     if (ec)
       throw std::runtime_error(stringify("filesystem error: ", ec));
 
-    llarp::LogInfo("Attempting to create config file, asRouter: ", asRouter, " path: ", confFile);
+    llarp::LogInfo(
+        "Attempting to create config file for ",
+        (asRouter ? "router" : "client"),
```
Comment (tewinget) [inline llarp/config/config.cpp:1142]: good call.

File: llarp/config/config.cpp [hunk 4]
```diff
@@ -1132,18 +1134,24 @@ namespace llarp
...
 
+    const auto parent = confFile.parent_path();
+
     // create parent dir if it doesn't exist
-    if (not fs::exists(confFile.parent_path(), ec))
+    if ((not parent.empty()) and (not fs::exists(parent, ec)))
     {
-      if (not fs::create_directory(confFile.parent_path()))
+      if (not fs::create_directory(parent))
         throw std::runtime_error(stringify("Failed to create parent directory for ", confFile));
```
Comment (tewinget) [inline llarp/config/config.cpp]: this should be something more like "Unable to create config file: unable to create directory \\"directory_path\\" and it does not exist."

**Review by tewinget — APPROVED** [review]

**Pull request merged**